src/gravel/controllers/kv.py

```diff
@@ -12,125 +12,395 @@
+                # bothering to try anything.  Same in every other method that
+                # does something with an ioctx.
+                # But we probably also need to look at a lock on it (because...
+                # we *know* this is a problem, or we just imagine it's a problem?)
 
-    async def get(self, key: str) -> Optional[str]:
-        """Get value for provided key"""
-        assert self._client
-        value, _ = await self._client.get(key)
+                with rados.WriteOpCtx() as op:
+                    self._ioctx.set_omap(op, (key,), (bvalue,))
+                    logger.debug("Doing write op")
+                    # write op naturally gets stuck if cluster badly degraded,
+                    # need to do this asynchronously somehow...?
+                    # TODO: what happens if it never succeeds?  The local
+                    # cache is newer than the cluster, then later, when
+                    # the cluster comes back, it will clobber the local
+                    # cache (this is where we're starting to need epochs
+                    # and probably re-inventing all sorts of cache and
+                    # filesystem logic that others have done before)
+                    # Possible solution:
+                    # - write async(?)
+                    # - write to local cache, but with "not known good" flag
+                    # - when async write completes, remove "not known good" flag
+                    #   from local cache value
+                    # - if async write never completes (how to detect?) then
+                    #   revert value to last known good (presupposes we're storing
+                    #   last known good values somehow as part of this flag
+                    #   business)
+                    # except that's still no good; ioctx.operate_aio_write_op()
+                    # *still* blocks if the cluster is sufficiently hosed -
+                    # following is test code to use to demonstrate this (it
+                    # hangs before either callback is hit)
+                    # def _write_complete(comp):
+                    #     logger.debug(f"write complete {comp}")
+                    # def _write_safe(comp):
+                    #     logger.debug(f"write safe {comp}")
+                    # ioctx.operate_aio_write_op(op, "kvstore",
+                    #     _write_complete,
+                    #     _write_safe)
+                    # Q: can we determine upfront whether a write will block?
+                    self._ioctx.operate_write_op(op, "kvstore")
+                    # This next notifies all watchers *INCLUDING* me!
+                    self._ioctx.notify("kvstore", key)
+            except Exception as e:
+                # TODO: deal with this (log it? ignore it?)
+                # e.g. RADOS rados state (You cannot perform that operation on a Rados object in state configuring.)
+                logger.error(str(e))
```

Comment:
(nit) if you use `logger.exception` it includes the stacktrace

Reply:
Done!